docs/tutorials/deployment.md

```diff
@@ -6,64 +6,74 @@ A few basic concepts about this process:
 __"Runtime"__ is an engine that loads a serialized model and executes it,
-e.g., PyTorch, Caffe2, TensorFlow, onnxruntime, TensorRT, etc.
+e.g., PyTorch, ONNX Runtime, TensorFlow, TensorRT, (DEPRECATED) Caffe2, etc.
 A runtime is often tied to a specific format
-(e.g. PyTorch needs TorchScript format, Caffe2 needs protobuf format).
+(e.g. PyTorch needs TorchScript format, ONNX Runtime requires ONNX format, (DEPRECATED) Caffe2 needs protobuf format).
 We currently support the following combination and each has some limitations:
 
 ```eval_rst
-+----------------------------+-------------+-------------+-----------------------------+
-|       Export Method        |   tracing   |  scripting  |       caffe2_tracing        |
-+============================+=============+=============+=============================+
-| **Formats**                | TorchScript | TorchScript | Caffe2, TorchScript, ONNX   |
-+----------------------------+-------------+-------------+-----------------------------+
-| **Runtime**                | PyTorch     | PyTorch     | Caffe2, PyTorch             |
-+----------------------------+-------------+-------------+-----------------------------+
-| C++/Python inference       | ✅          | ✅          | ✅                          |
-+----------------------------+-------------+-------------+-----------------------------+
-| Dynamic resolution         | ✅          | ✅          | ✅                          |
-+----------------------------+-------------+-------------+-----------------------------+
-| Batch size requirement     | Constant    | Dynamic     | Batch inference unsupported |
-+----------------------------+-------------+-------------+-----------------------------+
-| Extra runtime deps         | torchvision | torchvision | Caffe2 ops (usually already |
-|                            |             |             |                             |
-|                            |             |             | included in PyTorch)        |
-+----------------------------+-------------+-------------+-----------------------------+
-| Faster/Mask/Keypoint R-CNN | ✅          | ✅          | ✅                          |
-+----------------------------+-------------+-------------+-----------------------------+
-| RetinaNet                  | ✅          | ✅          | ✅                          |
-+----------------------------+-------------+-------------+-----------------------------+
-| PointRend R-CNN            | ✅          | ❌          | ❌                          |
-+----------------------------+-------------+-------------+-----------------------------+
-| Cascade R-CNN              | ✅          | ❌          | ❌                          |
-+----------------------------+-------------+-------------+-----------------------------+
++----------------------------+-------------+-------------+-------------------+-----------------------------+
+|       Export Method        |   tracing   |  scripting  | tracing/scripting | caffe2_tracing (DEPRECATED) |
```

Comment:
Unless this is implemented in `export_model.py` I don't think it's correct to way "scripting + ONNX" is supported.

Reply:
Not sure if I followed, in theory, wherever [scripted_module = torch.jit.script(model)` succeeds](https://github.com/facebookresearch/detectron2/blob/0e29b7ab8b77860b397aa02c2a164e1a8edcea8b/tests/test_export_torchscript.py), `torch.onnx.export(scripted_module)` succeeds by extension. The ONNX export does not need to do anything extra because of scripting or tracing.

Reply:
Does that actually work for scripting detectron2 models, though? The scripted model has complex IR including custom classes/list/dict/loops.
If it works, please add `--format=onnx` support to `export_scripting` in `export_model.py`.

Reply:
I just tried adding `torch.onnx.export(scripted_module, ...)` to `export_model.py:export_scripting()` and it does not work for me. Maybe I'm just using a bad pytorch version or I lack enough onnx knowledge.. but anyway please implement it first before claiming it in the docs.

Reply:
The error I had when exporting a maskrcnn with scripting+onnx:
```
Traceback (most recent call last):
  File "/home/USERNAME//detectron2/tools/deploy/./export_model.py", line 222, in <module>
    exported_model = export_scripting(torch_model)
  File "/home/USERNAME//detectron2/tools/deploy/./export_model.py", line 102, in export_scripting
    torch.onnx.export(ts_model, ({"image":image},), f, opset_version=STABLE_ONNX_OPSET_VERSION)
  File "/home/USERNAME/.local/lib/python3.10/site-packages/torch/onnx/utils.py", line 504, in export
    _export(
  File "/home/USERNAME/.local/lib/python3.10/site-packages/torch/onnx/utils.py", line 1506, in _export
    with exporter_context(model, training, verbose):
  File "/usr/lib/python3.10/contextlib.py", line 135, in __enter__
    return next(self.gen)
  File "/home/USERNAME/.local/lib/python3.10/site-packages/torch/onnx/utils.py", line 176, in exporter_context
    with select_model_mode_for_export(
  File "/usr/lib/python3.10/contextlib.py", line 135, in __enter__
    return next(self.gen)
  File "/home/USERNAME/.local/lib/python3.10/site-packages/torch/onnx/utils.py", line 116, in select_model_mode_for_export
    model.train(False)
  File "/home/USERNAME/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1905, in train
    self.training = mode
  File "/home/USERNAME/.local/lib/python3.10/site-packages/torch/jit/_script.py", line 794, in __setattr__
    self._c.setattr(attr, value)
RuntimeError: Can't set constant 'training' which has value:False
```

This looks related to one of the hacks we did to make scripting works.. I doubt there is an easy fix to this. Maybe just remove scripting from this PR. Tracing has been more useful and reliable anway.

Reply:
Tracing and scripting columns for ONNX were separated. Now it should be clearer that ONNX + tracing or ONNX + scripting didn't create any new limitation or capability that wasn't there for tracing or scripting alone

Reply:
ONNX + scripting doesn't seem to work according to my comment above.. in that case ONNX+scripting should be removed from doc

Reply:
please provide a full repro script

Reply:
```diff
diff --git i/tools/deploy/export_model.py w/tools/deploy/export_model.py
index f507dffe..ff991195 100755
--- i/tools/deploy/export_model.py
+++ w/tools/deploy/export_model.py
@@ -72,7 +72,6 @@ def export_scripting(torch_model):
         "pred_keypoints": torch.Tensor,
         "pred_keypoint_heatmaps": torch.Tensor,
     }
-    assert args.format == "torchscript", "Scripting only supports torchscript format."
 
     class ScriptableAdapterBase(nn.Module):
         # Use this adapter to workaround https://github.com/pytorch/pytorch/issues/46944
@@ -97,9 +96,10 @@ def export_scripting(torch_model):
                 return [i.get_fields() for i in instances]
 
     ts_model = scripting_with_instances(ScriptableAdapter(), fields)
-    with PathManager.open(os.path.join(args.output, "model.ts"), "wb") as f:
-        torch.jit.save(ts_model, f)
-    dump_torchscript_IR(ts_model, args.output)
+    if args.format == "onnx":
+        with PathManager.open(os.path.join(args.output, "model.onnx"), "wb") as f:
+            image = torch.ones((100, 100, 3))
+            torch.onnx.export(ts_model, ({"image":image},), f, opset_version=STABLE_ONNX_OPSET_VERSION)
     # TODO inference in Python now missing postprocessing glue code
     return None
 
```

```
./export_model.py --config-file ../../configs/quick_schedules/mask_rcnn_R_50_FPN_inference_acc_test.yaml --output ./mrcnn_scripting_onnx --export-method scripting --format onnx    MODEL.DEVICE cpu
```